main/JSONParameterUtilsMod.F90

```diff
@@ -0,0 +1,1304 @@
+             write(log_unit,*) 'encountered EOF'
+             call shr_sys_abort()
+          elseif (io_status < 0) then
+             cycle do_filepos
+          elseif (io_status>0) then
+             write(log_unit,*) 'fatal i/o error! code:', io_status
+             write(log_unit,*) 'msg: ',io_msg
+             call shr_sys_abort()
+          end if
+          
+          i=i+1
+          
+       end do do_filepos
+    end if
+    
+  end subroutine GotoPos
+
+  ! =====================================================================================
+  
+  subroutine StringToStringOrReal(string_in,out_string,out_real,is_num)
+
+    ! This routine takes a string, and determines if it is a number
+    ! by trying to read the string into a floating point number
+    ! It uses the iostat report to determine if it was possible or not
+    ! If the string was not a number, it returns a cleaned version of the
+    ! string. It was a number, it returns that as well.
+    
+    character(len=*),intent(in) :: string_in
+    logical,intent(in)          :: is_num  ! This specifies if we think the value is a number
+    character(len=max_sl),intent(out) :: out_string
+    real(r8),intent(out)              :: out_real
+
+    character(len=max_sl)       :: tmp_str
+    character(len=max_sl)       :: tmp_lc_str
+    integer  :: io_status
+    real(r8) :: tmp_real
+    logical  :: is_invalid
+    
+    tmp_str = trim(CleanSymbol(string_in))
+
+    tmp_lc_str = tmp_str
+
+    call to_lowercase(tmp_lc_str)
+    
+    ! Check if this is a nan
+    is_invalid = index(tmp_lc_str,'nan')>0 .or. index(tmp_lc_str,'null')>0
+
+    if(is_invalid)then
+       out_real = r_invalid
+       out_string = 'null'
+    else
+       read (unit=tmp_str,fmt=*,iostat=io_status) tmp_real
+       if (io_status == 0) then
+          if(is_num)then
+             out_real = tmp_real
+             out_string = 'real number'
+          else
+             write(log_unit,*) 'Read string, expected number',trim(CleanSymbol(string_in))
+             call shr_sys_abort()
+          end if
+       elseif(io_status /= 0) then
+          if(.not.is_num)then
+             out_real = -9999.9_r8
+             out_string = trim(CleanSymbol(string_in))
+          else
+             write(log_unit,*) 'Read string, expected character',trim(CleanSymbol(string_in))
+             call shr_sys_abort()
+          end if
+       end if
+    end if
+    return
+  end subroutine StringToStringOrReal
+
+  ! =====================================================================================
+
+  subroutine to_lowercase(input_string)
+    character(len=*), intent(inout) :: input_string
+    integer :: i, char_code
+
+    ! Use IACHAR to convert the character to its integer
+    ! code. Capital letters have a different code than lower
+    ! case.
+    
+    do i = 1,len(input_string)
+
+       char_code = iachar(input_string(i:i))
+        
+       ! Check if the character is an uppercase letter (A-Z)
```

Comment:
This does rely on ASCII, which if it can get the ASCII representation will function. But, won't if not. I think iachar returns 0 when it's not in ASCII. So you should add an error when it's zero. So this implicitly adds a requirement that input JSON files are ASCII representable, which is probably fine. But, something we should increase visibility for.

I did find an example function that doesn't require ASCII at the end of this stackloverflow answer.

https://stackoverflow.com/questions/10759375/how-can-i-write-a-to-upper-or-to-lower-function-in-f90

But, it still might be OK to insist that characters in our JSON files are ASCII representable. Something to think about.